### PR TITLE
Add plugin helper

### PR DESCRIPTION
**Pull request by achambers**
## Add `pluginHelper` to config property functions

In order to access the default value of a plugin property from within a consumer defined config function, a `pluginHelper` is now passed in like so:

``` javascript
ENV['foo'] = {
  names: function(context, pluginHelper) {
    return pluginHelper.readConfigDefault('distFiles').push('index.json'); #=> ['index.htm', 'index.json']
  }
};
```

## Fix `this` context inside config property functions

If I have a `deploy.js` config that does the following:

``` javascript
// config/deploy.js

ENV.foo = {
  bar: function() {
    console.log(this);
  }
}
```

Now, if the `ember-cli-deploy-foo` plugin extends this base plugin, imagine the following code:

``` javascript
upload: function(context) {
  context.foo.bar();
  this.readConfig('bar');
}
```

In this example, in the first call (`context.foo.bar()`), as you would expect, `this` is the containing object - `ENV.foo`. However, in the second call (`this.readConfig('bar')`), `this` is actually the base plugin.

This can lead to confusion and error as `this` will be different based on how it is accessed.

This PR fixes this so that `this` remains the object containing the function which is what you would expect as a consumer of the plugin.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -34,6 +34,16 @@ module.exports = {
       },
       requiredConfig: ['awesomeApiKey'], // throw an error if this is not configured
 
+      // return an object from here to add functionality to the `pluginHelper`
+      // that is passed in to user defined config property functions
+      pluginHelper: function(context) {
+        return {
+          sayHello: function() {
+            return 'Hello' + context.project.name();
+          }
```
Comment (lukemelia) [inline README.md]: Can we come with a somewhat plausible example from the deployment domain here?

File: index.js [hunk 2]
```diff
@@ -1,5 +1,18 @@
 var CoreObject     = require('core-object');
 var chalk = require('chalk');
+var merge = require('lodash.merge');
+
+function _createPluginHelper() {
```
Comment (lukemelia) [inline index.js]: naming of this function seems off. Maybe it is `_pluginHelperDefaults`?

File: README.md (hunk 1 above)
Comment (achambers) [reply README.md]: Yep. Was low on inspiration this evening. Did you have any thoughts on how people might use this but of functionality? Can't really think of how I might use this in the real world to be honest.

File: index.js (hunk 2 above)
Comment (achambers) [reply index.js]: Ah yeh. That naming is a hangover from how I wrote it initially. Will change now

File: README.md (hunk 1 above)
Comment (lukemelia) [reply README.md]: The example that sold me on this was adding an item to a list of file extensions.

Comment (achambers) [reply README.md]: This is different right? This is the functionality that suggested that plugin author could add to the pluginHelper.

Comment (achambers) [reply README.md]: I'd actually be happy to remove this for now until we find a use case as I cent think of how I'd use it in real life.

Comment (achambers) [reply README.md]: Massive fails at English. Typing on phone. Sorry

Comment (lukemelia) [reply README.md]: Ah, right. OK, I can't think of one either.

Comment (achambers) [reply README.md]: Ok then. So let's remove it and wait and see how it's used and see if some use cases emerge.

File: tests/unit/index-nodetest.js [hunk 3]
```diff
@@ -70,7 +70,68 @@ describe('base plugin', function() {
...
+
+      var context = {
+        foo: 'foo',
+        config: {
+          blah: {
+            foo: function(context, pluginHelper) {
+              return pluginHelper.readConfigDefault('foo') + 'foo';
+            },
+            bar: function(context, pluginHelper) {
+              return pluginHelper.readConfigDefault('bar') + 'bar';
```
Comment (LevelbossMike) [inline tests/unit/index-nodetest.js]: as bar is not defined prior on the default configuration do we want to make sure that this does not fail and returns `undefined` as expected? when reading this the first time it looked kind of strange to check for a string of "undefinedbar".

It might also be useful to add a test that static values on the default configuration work as expected as well.

File: tests/unit/index-nodetest.js [hunk 4]
```diff
@@ -70,7 +70,68 @@ describe('base plugin', function() {
...
+          return { bar: context.foo };
+        }
+      });
+
+      var context = {
+        foo: 'bar',
+        config: {
+          blah: {
+            foo: function(context, pluginHelper) {
+              return pluginHelper.readConfigDefault('foo') + pluginHelper.bar;
```
Comment (LevelbossMike) [inline tests/unit/index-nodetest.js]: As all relevant variables/properties of this test are called `foo` and `bar` I found it pretty hard to follow what's going on in this test. This may be just me but imo it would be easier to follow if the property on the context that can be accessed via the property configured on pluginHelper by the implementor would be called differently than the default configuration option we access here via `pluginHelper.readConfigDefault`

File: README.md (hunk 1 above)
Comment (LevelbossMike) [reply README.md]: maybe something that gets added via revisions-data would be a useful example here. I assume notification plugins like slack or similiar could be interested in information that is easily accessible via helper functions like "deployerName" or "version".

But the project name itself is a useful example for notification plugins in itself when thinking about it. :+1: maybe just return a property called `projectName` and return just `context.project.name()` itself.

**Comment by achambers** [comment]
Thanks for the feedback @levelbossmike. Will go through the comments and action accordingly.

**Comment by achambers** [comment]
@LevelbossMike / @lukemelia / @ghedamat I have finally gotten back to this PR. I removed the ability for plugin authors to add functions to the `pluginHelper` as we don't really have a use case for it right now so better to leave it out until it is proven to be needed. As a result, tests have been simplified.

Please review and merge as appropriate.

**Comment by LevelbossMike** [comment]
🎉
